linker/src/bus.rs

```diff
@@ -0,0 +1,354 @@
+                parse_pil_statement(&format!(
+                    "std::array::map({call_selectors}, std::utils::force_bool);"
+                )),
+            ])
+        }
+
+        // process outgoing links
+        for link in &object.links {
+            self.process_link(link, location, objects);
+        }
+        // receive incoming links
+        for (name, operation) in &object.operations {
+            self.process_operation(name, operation, location, object);
+        }
+
+        // if this is the main object, call the main operation
+        if *location == Location::main() {
+            let operation_id = object.operation_id.clone();
+            let main_operation_id = object
+                .operations
+                .get(MAIN_OPERATION_NAME)
+                .and_then(|operation| operation.id.as_ref());
+
+            if let (Some(operation_id_name), Some(operation_id_value)) =
+                (operation_id, main_operation_id)
+            {
+                self.pil
+                    .extend(call(&operation_id_name, operation_id_value));
+            }
+        }
+    }
+
+    fn into_pil(self) -> Vec<PilStatement> {
+        self.pil
+    }
+}
+
+impl BusLinker {
+    /// Process an operation, which means receive from the bus with the correct [InteractionType]
+    fn process_operation(
+        &mut self,
+        operation_name: &str,
+        operation: &Operation,
+        location: &Location,
+        object: &Object,
+    ) {
+        let link_to = LinkTo {
+            machine: location.clone(),
+            operation: operation_name.to_string(),
+        };
+
+        let selector_index = self.selector_array_index_by_operation.get(&link_to);
+
+        // By construction, all operations *which are called* have an optional selector index. The others can be safely ignored.
+        if let Some(selector_index) = selector_index {
+            // compute the unique interaction id
+            let interaction_id = interaction_id(&link_to);
+
+            let namespace = location.to_string();
+
+            let op_id = object
+                .operation_id
+                .clone()
+                .map(|oid| namespaced_reference(namespace.clone(), oid))
+                .into_iter();
+
+            let tuple = ArrayLiteral {
+                items: op_id
+                    .chain(operation.params.inputs_and_outputs().map(|i| {
+                        index_access(
+                            namespaced_reference(namespace.clone(), &i.name),
+                            i.index.clone(),
+                        )
+                    }))
+                    .collect(),
+            }
+            .into();
+
+            let latch = namespaced_reference(namespace.clone(), object.latch.clone().unwrap());
+
+            let (function, arguments) = match selector_index {
+                // a selector index of None means this operation is called via lookup
```

Comment:
Is this really always an invariant? I mean, makes sense but just wondering.

Reply:
yes, when we construct `selector_array_index_by_operation` we insert None for lookups and Some(index) for permutations.

Note that `self.selector_array_index_by_operation.get` returns `Option<Option<index>>>`, the outer one to spot unused operations (for which we don't receive), the inner one to distinguish operations accessed via lookup vs permutation.

Reply:
This could be cleaner if we had an optimizer pass on the object graph removing unused operations. We could then unwrap the `get` call.

Reply:
ok